### DISCUSSION TITLE
Testcontainers with MSTest

### DISCUSSION THREAD

**Discussion by jc1231** [testcontainers/testcontainers-dotnet#1170] [Q&A] [answered]
I cannot find any articles or examples or best practices using Testcontainers with MSTest. Everything appears to be for XUnit only.

`IAsyncLifetime` for example doesn't seem to have an equivalent in MSTest.

Are there any examples or guides or is it generally not supported?

**Comment by HofmeisterAn** [comment, marked as answer]
The equivalent implementation in [MSTest](https://learn.microsoft.com/en-us/visualstudio/test/using-microsoft-visualstudio-testtools-unittesting-members-in-unit-tests#test) will utilize the `[TestInitialize]` and `[TestCleanup]` attributes (annotations). Please be aware and implement the async/await pattern properly, e.g.:

```csharp
[TestClass]
public class MyTestClass
{
    private readonly PostgreSqlContainer _postgreSqlContainer = new PostgreSqlBuilder().Build();

    [TestInitialize]
    public Task TestInitialize()
    {
        return _postgreSqlContainer.StartAsync();
    }

    [TestCleanup]
    public Task TestCleanup()
    {
        return _postgreSqlContainer.DisposeAsync().AsTask();
    }
}
```

There is nothing special. It works the same way as you would set up any other fixture in MSTest as well.

**Comment by jc1231** [comment]
@HofmeisterAn Thanks for the info this is helpful.

The other issue I'm trying to solve is performance. I've seen many articles on using a single test container per test collection vs one for every individual test. XUnit appears to solve this with something called collection fixtures, I could not find an equivalent for MSTest.
MSTest appears to run everything in parallel even when my test container is a singleton.

Another strange thing I've noticed is even when running a single test, 7+ db containers will get spun up.

**Reply by HofmeisterAn** [reply]
This depends on your chosen test framework. For MSTest, please familiarize yourself with its test lifecycle attributes. I think this [article](https://www.meziantou.net/mstest-v2-test-lifecycle-attributes.htm) summarizes it well.
